File: src/tgls_new.c
#include <stdio.h>

#include <caml/alloc.h>
#include <caml/custom.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>
#include <caml/fail.h>
#include <caml/bigarray.h>

#include <OpenGL/Gl.h>

// CAMLprim value TglClearColor(value r, value g, value b, value a) {
//   CAMLparam4(r, g, b, a);
//   glClearColor();
//   CAMLreturn0;
// }

CAMLprim value TglCreateProgram() {
  CAMLparam0();
  CAMLreturn(Val_int(glCreateProgram()));
}

CAMLprim value TglCreateShader(value shaderType) {
  CAMLparam1(shaderType);
  CAMLreturn(Val_int(glCreateShader(Int_val(shaderType))));
}

void TglAttachShader(value program, value shader) {
  CAMLparam2(program, shader);
  glAttachShader(Int_val(program), Int_val(shader));
  CAMLreturn0;
}

void TglDeleteShader(value shader) {
  CAMLparam1(shader);
  glDeleteShader(Int_val(shader));
  CAMLreturn0;
}

void TglShaderSource(value shader, value stringArray) {
  CAMLparam2(shader, stringArray);
  int numOfElements = Wosize_val(stringArray);
  const GLchar **arrOfElements = malloc(sizeof(GLchar *) * numOfElements);
  GLint *arrOfLengths = malloc(sizeof(GLint) * numOfElements);
  for(int i = 0; i < numOfElements; ++i) {
    value e = Field(stringArray, i);
    arrOfElements[i] = String_val(e);
    arrOfLengths[i] = caml_string_length(e);
  }
  glShaderSource(Int_val(shader), numOfElements, arrOfElements, arrOfLengths);
  CAMLreturn0;
}

void TglCompileShader(value shader) {
  CAMLparam1(shader);
  glCompileShader(Int_val(shader));
  CAMLreturn0;
}

void TglLinkProgram(value program) {
  CAMLparam1(program);
  glLinkProgram(Int_val(program));
  CAMLreturn0;
}

void TglUseProgram(value program) {
  CAMLparam1(program);
  glUseProgram(Int_val(program));
  CAMLreturn0;
}

CAMLprim value TglGenBuffers(value count) {
  CAMLparam1(count);
  CAMLlocal1(ret);

  int size = Int_val(count);
  unsigned int *buffers = malloc(sizeof(unsigned int) * size);
  glGenBuffers(size, buffers);

  ret = caml_alloc_small(size, 0);
  for (int i = 0; i < size; ++i) {
    Field(ret, i) = Val_int(buffers[i]);
  }
  CAMLreturn(ret);
}

CAMLprim value TglGenBuffer() {
  CAMLparam0();
  unsigned int buffers = 0;
  glGenBuffers(1, &buffers);
  CAMLreturn(Val_int(buffers));
}

void TglClearColor(value r, value g, value b, value a) {
  CAMLparam4(r, g, b, a);
  glClearColor(Double_val(r), Double_val(g), Double_val(b), Double_val(a));
  CAMLreturn0;
}

void TglBindBuffer(value kind, value buffer) {
  CAMLparam2(kind, buffer);
  glBindBuffer(Int_val(kind), Int_val(buffer));
  CAMLreturn0;
}

CAMLprim value TglGenTextures(value count) {
  CAMLparam1(count);
  CAMLlocal1(ret);

  int size = Int_val(count);
  unsigned int *textures = malloc(sizeof(unsigned int) * size);
  glGenTextures(size, textures);

  ret = caml_alloc_small(size, 0);
  for (int i = 0; i < size; ++i) {
    Field(ret, i) = Val_int(textures[i]);
  }

  CAMLreturn(ret);
}

CAMLprim value TglGenTexture() {
  CAMLparam0();
  unsigned int textures = 0;
  glGenTextures(1, &textures);
  CAMLreturn(Val_int(textures));
}

void TglActiveTexture(value textureUnit) {
  CAMLparam1(textureUnit);
  glActiveTexture(Int_val(textureUnit));
  CAMLreturn0;
}

void TglBindTexture(value kind, value texture) {
  CAMLparam2(kind, texture);
  glBindTexture(Int_val(kind), Int_val(texture));
  CAMLreturn0;
}

void TglTexSubImage2D_native(value target, value level, value xoffset, value yoffset, value width, value height, value format, value type, value pixels) {
  CAMLparam5(target, level, xoffset, yoffset, width);
  CAMLxparam4(height, format, type, pixels);
  
  glTexSubImage2D(Int_val(target), Int_val(level), Int_val(xoffset), Int_val(yoffset), Int_val(width), Int_val(height), Int_val(format), Int_val(type), Caml_ba_data_val(pixels));
  CAMLreturn0;
}

void TglTexSubImage2D_bytecode(value * argv, int argn) {
  TglTexSubImage2D_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5], argv[6], argv[7], argv[8]);
}


void TglTexParameteri(value kind, value pname, value param) {
  CAMLparam3(kind, pname, param);
  glTexParameteri(Int_val(kind), Int_val(pname), Int_val(param));
  CAMLreturn0;
}

void TglEnable(value thing) {
  CAMLparam1(thing);
  glEnable(Int_val(thing));
  CAMLreturn0;
}

void TglDisable(value thing) {
  CAMLparam1(thing);
  glDisable(Int_val(thing));
  CAMLreturn0;
}

void TglBlendFunc(value sfactor, value dfactor) {
  CAMLparam2(sfactor, dfactor);
  glBlendFunc(Int_val(sfactor), Int_val(dfactor));
  CAMLreturn0;
}

CAMLprim value TglReadPixels_RGBA(value x, value y, value width, value height) {
  CAMLparam4(x, y, width, height);
  CAMLlocal1(ret);

  // Allocate a pointer for caml_ba_alloc's sake.
  intnat *size = malloc(sizeof(intnat));
  *size = Int_val(width) * Int_val(height) * 4;

  char *data = malloc(*size * sizeof(char));

  glReadPixels(Int_val(x), Int_val(y), Int_val(width), Int_val(height), GL_RGBA, GL_UNSIGNED_BYTE, data);

  // return array of size `size` of dimension 1 of uint8 (char).
  ret = caml_ba_alloc(CAML_BA_UINT8, 1, data, size);
  CAMLreturn(ret);
}

void TglTexImage2D_RGBA_native(value target, value level, value width, value height, value border, value data) {
  CAMLparam5(target, level, width, height, border);
  CAMLxparam1(data);
  glTexImage2D(Int_val(target), Int_val(level), 4, Int_val(width), Int_val(height), Int_val(border), GL_RGBA, GL_UNSIGNED_BYTE, Caml_ba_data_val(data));
  CAMLreturn0;
}

void TglTexImage2D_RGBA_bytecode(value * argv, int argn) {
  TglTexImage2D_RGBA_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

void TglUniform1i(value location, value v) {
  CAMLparam2(location, v);
  glUniform1i(Int_val(location), Int_val(v));
  CAMLreturn0;
}

void TglUniform1f(value location, value v) {
  CAMLparam2(location, v);
  glUniform1f(Int_val(location), Double_val(v));
  CAMLreturn0;
}

void TglBufferData(value target, value data, value usage) {
  CAMLparam3(target, data, usage);
  glBufferData(Int_val(target), caml_ba_byte_size(Caml_ba_array_val(data)), Caml_ba_data_val(data), Int_val(usage));
  CAMLreturn0;
}

void TglViewport(value x, value y, value width, value height) {
  CAMLparam4(x, y, width, height);
  glViewport(Int_val(x), Int_val(y), Int_val(width), Int_val(height));
  CAMLreturn0;
}

void TglClear(value mask) {
  CAMLparam1(mask);
  glClear(Int_val(mask));
  CAMLreturn0;
}

CAMLprim value TglGetUniformLocation(value program, value name) {
  CAMLparam2(program, name);
  CAMLreturn(Val_int(glGetUniformLocation(Int_val(program), String_val(name))));
}

CAMLprim value TglGetAttribLocation(value program, value name) {
  CAMLparam2(program, name);
  CAMLreturn(Val_int(glGetAttribLocation(Int_val(program), String_val(name))));
}

void TglEnableVertexAttribArray(value attrib) {
  CAMLparam1(attrib);
  glEnableVertexAttribArray(Int_val(attrib));
  CAMLreturn0;
}

void TglVertexAttribPointer_native(value index, value size, value typ, value normalize, value stride, value offset) {
  CAMLparam5(index, size, typ, normalize, stride);
  CAMLxparam1(offset);
  long o = (long)Int_val(offset);
  glVertexAttribPointer(Int_val(index), Int_val(size), Int_val(typ), Bool_val(normalize), Int_val(stride), (const GLvoid *)o);
  CAMLreturn0;
}

void TglVertexAttribPointer_bytecode(value * argv, int argn) {
  TglVertexAttribPointer_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

CAMLprim value TglGetProgramiv(value program, value pname) {
  CAMLparam2(program, pname);
  GLint ret;
  glGetProgramiv(Int_val(program), Int_val(pname), &ret);
  CAMLreturn(Val_int(ret));
}

CAMLprim value TglGetShaderiv(value shader, value pname) {
  CAMLparam2(shader, pname);
  GLint ret;
  glGetShaderiv(Int_val(shader), Int_val(pname), &ret);
  CAMLreturn(Val_int(ret));
}

CAMLprim value TglGetShaderInfoLog(value shader) {
  CAMLparam1(shader);
  CAMLlocal1(ret);

  GLint exactLength;
  glGetShaderiv(Int_val(shader), GL_INFO_LOG_LENGTH, &exactLength);
  GLchar *buffer = malloc(exactLength * sizeof(char));
  glGetShaderInfoLog(Int_val(shader), exactLength - 1, NULL, buffer);

  ret = caml_copy_string(buffer);
  CAMLreturn(ret);
}

CAMLprim value TglGetProgramInfoLog(value program) {
  CAMLparam1(program);
  CAMLlocal1(ret);

  GLint exactLength;
  glGetProgramiv(Int_val(program), GL_INFO_LOG_LENGTH, &exactLength);

  GLchar *buffer = malloc(exactLength * sizeof(char));
  glGetProgramInfoLog(Int_val(program), exactLength - 1, NULL, buffer);
  ret = caml_copy_string(buffer);
  CAMLreturn(ret);
}

CAMLprim value TglGetShaderSource(value shader) {
  CAMLparam1(shader);
  CAMLlocal1(ret);

  GLint exactLength;
  glGetShaderiv(Int_val(shader), GL_SHADER_SOURCE_LENGTH, &exactLength);

  GLchar *buffer = malloc(exactLength * sizeof(char));
  glGetShaderSource(Int_val(shader), exactLength - 1, NULL, buffer);
  ret = caml_copy_string(buffer);
  CAMLreturn(ret);
}

void TglDrawArrays(value mode, value first, value count) {
  CAMLparam3(mode, first, count);
  glDrawArrays(Int_val(mode), Int_val(first), Int_val(count));
  CAMLreturn0;
}

void TglDrawElements(value mode, value first, value typ, value offset) {
  CAMLparam4(mode, first, typ, offset);
  long o = (long)Int_val(offset);
  glDrawElements(Int_val(mode), Int_val(first), Int_val(typ), (const GLvoid *)o);
  CAMLreturn0;
}

void TglUniformMatrix4fv(value location, value transpose, value val) {
  CAMLparam3(location, transpose, val);
  int size = Wosize_val(val);
  float *matrix = malloc(sizeof(float) * size);
  for (int i = 0; i < size; ++i){
    matrix[i] = Double_field(val, i);
  }
  glUniformMatrix4fv(Int_val(location), 1, Bool_val(transpose), matrix);
  CAMLreturn0;
}

File: src/tgls_new.ml
#if BYTECODE || NATIVE then

external gladLoadGL : unit -> unit = "gladLoadGL"[@@noalloc ]
type programT
external createProgram : unit -> programT = "TglCreateProgram"[@@noalloc ]
type shaderT
let gl_vertex_shader = 35633
let gl_fragment_shader = 35632
external createShader : int -> shaderT = "TglCreateShader"[@@noalloc ]
external attachShader :
  program:programT -> shader:shaderT -> unit = "TglAttachShader"[@@noalloc ]
external deleteShader : shaderT -> unit = "TglDeleteShader"[@@noalloc ]
external shaderSource : shaderT -> string array -> unit = "TglShaderSource"
external compileShader : shaderT -> unit = "TglCompileShader"[@@noalloc ]
external linkProgram : programT -> unit = "TglLinkProgram"[@@noalloc ]
external useProgram : programT -> unit = "TglUseProgram"[@@noalloc ]
type bufferT
let gl_array_buffer = 34962
let gl_element_array_buffer = 34963
let gl_pixel_pack_buffer = 35051
let gl_pixel_unpack_buffer = 35052
external bindBuffer : target:int -> buffer:bufferT -> unit = "TglBindBuffer"
[@@noalloc ]
external genBuffers : int -> bufferT array = "TglGenBuffers"
external genBuffer : unit -> bufferT = "TglGenBuffer"[@@noalloc ]
external clearColor :
  red:float -> green:float -> blue:float -> alpha:float -> unit =
    "TglClearColor"[@@noalloc ]
type textureT
external genTextures : int -> textureT array = "TglGenTextures"
external genTexture : unit -> textureT = "TglGenTexture"[@@noalloc ]
type framebufferT
external genFramebuffers : int -> framebufferT array = "TglGenFramebuffers"
external genFramebuffer : unit -> framebufferT = "TglGenFramebuffer"[@@noalloc
                                                                    ]
let gl_framebuffer = 36160
let gl_color_attachment0 = 36064
external bindFramebuffer : int -> framebufferT -> unit = "TglBindFramebuffer"
[@@noalloc ]
external bindDefaultFramebuffer : int -> unit = "TglBindDefaultFramebuffer"
[@@noalloc ]
external framebufferTexture2D :
  target:int ->
    attachment:int -> texTarget:int -> texture:textureT -> level:int -> unit
    = "TglFramebufferTexture2D"[@@noalloc ]
let gl_texture0 = 33984
let gl_texture1 = 33985
let gl_texture2 = 33986
let gl_texture3 = 33987
let gl_texture4 = 33988
let gl_texture5 = 33989
let gl_texture6 = 33990
let gl_texture7 = 33991
let gl_texture8 = 33992
let gl_texture9 = 33993
let gl_texture10 = 33994
let gl_texture11 = 33995
let gl_texture12 = 33996
let gl_texture13 = 33997
let gl_texture14 = 33998
let gl_texture15 = 33999
let gl_texture16 = 34000
let gl_texture17 = 34001
let gl_texture18 = 34002
let gl_texture19 = 34003
let gl_texture20 = 34004
let gl_texture21 = 34005
let gl_texture22 = 34006
let gl_texture23 = 34007
let gl_texture24 = 34008
let gl_texture25 = 34009
let gl_texture26 = 34010
let gl_texture27 = 34011
let gl_texture28 = 34012
let gl_texture29 = 34013
let gl_texture30 = 34014
let gl_texture31 = 34015
external activeTexture : int -> unit = "TglActiveTexture"[@@noalloc ]
let gl_texture_1d = 3552
let gl_texture_2d = 3553
let gl_texture_3d = 32879
let gl_texture_cube_map = 34067
external bindTexture :
  target:int -> texture:textureT -> unit = "TglBindTexture"[@@noalloc ]
external texSubImage2D :
  target:int ->
    level:int ->
      xoffset:int ->
        yoffset:int ->
          width:int ->
            height:int ->
              format:int ->
                type_:int ->
                  pixels:('a, 'b, Bigarray.c_layout) Bigarray.Array1.t ->
                    unit = "TglTexSubImage2D_bytecode"
    "TglTexSubImage2D_native"
let gl_texture_mag_filter = 10240
let gl_texture_min_filter = 10241
let gl_texture_wrap_s = 10242
let gl_texture_wrap_t = 10243
let gl_texture_min_lod = 33082
let gl_texture_max_lod = 33083
let gl_texture_base_level = 33084
let gl_texture_max_level = 33085
let gl_texture_wrap_r = 32882
let gl_texture_priority = 32870
let gl_texture_compare_mode = 34892
let gl_texture_compare_func = 34893
let gl_depth_texture_mode = 34891
let gl_generate_mipmap = 33169
external texParameteri :
  target:int -> pname:int -> param:int -> unit = "TglTexParameteri"[@@noalloc
                                                                    ]
let gl_alpha_test = 3008
let gl_auto_normal = 3456
let gl_blend = 3042
let gl_clip_plane0 = 12288
let gl_clip_plane1 = 12289
let gl_clip_plane2 = 12290
let gl_clip_plane3 = 12291
let gl_clip_plane4 = 12292
let gl_clip_plane5 = 12293
let gl_color_logic_op = 3058
let gl_color_material = 2903
let gl_color_sum = 33880
let gl_color_table = 32976
let gl_convolution_1d = 32784
let gl_convolution_2d = 32785
let gl_cull_face = 2884
let gl_depth_test = 2929
let gl_dither = 3024
let gl_fog = 2912
let gl_histogram = 32804
let gl_index_logic_op = 3057
let gl_light0 = 16384
let gl_light1 = 16385
let gl_light2 = 16386
let gl_light3 = 16387
let gl_light4 = 16388
let gl_light5 = 16389
let gl_light6 = 16390
let gl_light7 = 16391
let gl_lighting = 2896
let gl_line_smooth = 2848
let gl_line_stipple = 2852
let gl_map1_color_4 = 3472
let gl_map1_index = 3473
let gl_map1_normal = 3474
let gl_map1_texture_coord_1 = 3475
let gl_map1_texture_coord_2 = 3476
let gl_map1_texture_coord_3 = 3477
let gl_map1_texture_coord_4 = 3478
let gl_map1_vertex_3 = 3479
let gl_map1_vertex_4 = 3480
let gl_map2_color_4 = 3504
let gl_map2_index = 3505
let gl_map2_normal = 3506
let gl_map2_texture_coord_1 = 3507
let gl_map2_texture_coord_2 = 3508
let gl_map2_texture_coord_3 = 3509
let gl_map2_texture_coord_4 = 3510
let gl_map2_vertex_3 = 3511
let gl_map2_vertex_4 = 3512
let gl_minmax = 32814
let gl_multisample = 32925
let gl_normalize = 2977
let gl_point_smooth = 2832
let gl_point_sprite = 34913
let gl_polygon_offset_fill = 32823
let gl_polygon_offset_line = 10754
let gl_polygon_offset_point = 10753
let gl_polygon_smooth = 2881
let gl_polygon_stipple = 2882
let gl_post_color_matrix_color_table = 32978
let gl_post_convolution_color_table = 32977
let gl_rescale_normal = 32826
let gl_sample_alpha_to_coverage = 32926
let gl_sample_alpha_to_one = 32927
let gl_sample_coverage = 32928
let gl_separable_2d = 32786
let gl_scissor_test = 3089
let gl_stencil_test = 2960
let gl_texture_gen_s = 3168
let gl_texture_gen_t = 3169
let gl_texture_gen_r = 3170
let gl_texture_gen_q = 3171
let gl_vertex_program_point_size = 34370
let gl_vertex_program_two_side = 34371
external enable : int -> unit = "TglEnable"[@@noalloc ]
external disable : int -> unit = "TglDisable"[@@noalloc ]
let gl_zero = 0
let gl_one = 1
let gl_src_color = 768
let gl_one_minus_src_color = 769
let gl_src_alpha = 770
let gl_one_minus_src_alpha = 771
let gl_dst_alpha = 772
let gl_one_minus_dst_alpha = 773
let gl_dst_color = 774
let gl_one_minus_dst_color = 775
let gl_src_alpha_saturate = 776
let gl_constant_color = 32769
let gl_one_minus_constant_color = 32770
let gl_constant_alpha = 32771
let gl_one_minus_constant_alpha = 32772
external blendFunc : sfactor:int -> dfactor:int -> unit = "TglBlendFunc"
[@@noalloc ]
external readPixels_RGBA :
  x:int ->
    y:int ->
      width:int ->
        height:int ->
          (int, Bigarray.int8_unsigned_elt, Bigarray.c_layout)
            Bigarray.Array1.t = "TglReadPixels_RGBA"
let gl_proxy_texture_2d = 32868
let gl_texture_cube_map_positive_x = 34069
let gl_texture_cube_map_negative_x = 34070
let gl_texture_cube_map_positive_y = 34071
let gl_texture_cube_map_negative_y = 34072
let gl_texture_cube_map_positive_z = 34073
let gl_texture_cube_map_negative_z = 34074
let gl_proxy_texture_cube_map = 34075
let gl_alpha = 6406
let gl_alpha4 = 32827
let gl_alpha8 = 32828
let gl_alpha12 = 32829
let gl_alpha16 = 32830
let gl_compressed_alpha = 34025
let gl_compressed_luminance = 34026
let gl_compressed_luminance_alpha = 34027
let gl_compressed_intensity = 34028
let gl_compressed_rgb = 34029
let gl_compressed_rgba = 34030
let gl_depth_component = 6402
let gl_depth_component16 = 33189
let gl_depth_component24 = 33190
let gl_depth_component32 = 33191
let gl_luminance = 6409
let gl_luminance4 = 32831
let gl_luminance8 = 32832
let gl_luminance12 = 32833
let gl_luminance16 = 32834
let gl_luminance4_alpha4 = 32835
let gl_luminance6_alpha2 = 32836
let gl_luminance8_alpha8 = 32837
let gl_luminance12_alpha4 = 32838
let gl_luminance12_alpha12 = 32839
let gl_luminance16_alpha16 = 32840
let gl_intensity = 32841
let gl_intensity4 = 32842
let gl_intensity8 = 32843
let gl_intensity12 = 32844
let gl_intensity16 = 32845
let gl_r3_g3_b2 = 10768
let gl_rgb4 = 32847
let gl_rgb5 = 32848
let gl_rgb8 = 32849
let gl_rgb10 = 32850
let gl_rgb12 = 32851
let gl_rgb16 = 32852
let gl_rgba2 = 32853
let gl_rgba4 = 32854
let gl_rgb5_a1 = 32855
let gl_rgba8 = 32856
let gl_rgb10_a2 = 32857
let gl_rgba12 = 32858
let gl_rgba16 = 32859
let gl_sluminance_alpha = 35908
let gl_sluminance8_alpha8 = 35909
let gl_sluminance = 35910
let gl_sluminance8 = 35911
let gl_srgb = 35904
let gl_srgb8 = 35905
let gl_srgb_alpha = 35906
let gl_srgb8_alpha8 = 35907
external texImage2D_null :
  target:int -> level:int -> width:int -> height:int -> border:int -> unit =
    "TglTexImage2D_null"[@@noalloc ]
external texImage2D_RGBA :
  target:int ->
    level:int ->
      width:int ->
        height:int ->
          border:int ->
            data:('a, 'b, Bigarray.c_layout) Bigarray.Array1.t -> unit =
    "TglTexImage2D_RGBA_bytecode" "TglTexImage2D_RGBA_native"[@@noalloc ]
type uniformT
external uniform1i : location:uniformT -> value:int -> unit = "TglUniform1i"
[@@noalloc ]
external uniform1f :
  location:uniformT -> value:float -> unit = "TglUniform1f"[@@noalloc ]
external uniform2f :
  location:uniformT -> v1:float -> v2:float -> unit = "TglUniform2f"[@@noalloc
                                                                    ]
external uniform3f :
  location:uniformT -> v1:float -> v2:float -> v3:float -> unit =
    "TglUniform3f"[@@noalloc ]
external uniform4f :
  location:uniformT -> v1:float -> v2:float -> v3:float -> v4:float -> unit =
    "TglUniform4f"[@@noalloc ]
let gl_stream_draw = 35040
let gl_stream_read = 35041
let gl_stream_copy = 35042
let gl_static_draw = 35044
let gl_static_read = 35045
let gl_static_copy = 35046
let gl_dynamic_draw = 35048
let gl_dynamic_read = 35049
let gl_dynamic_copy = 35050
external bufferData :
  target:int ->
    data:('a, 'b, Bigarray.c_layout) Bigarray.Array1.t ->
      usage:int -> unit = "TglBufferData"[@@noalloc ]
external viewport :
  x:int -> y:int -> width:int -> height:int -> unit = "TglViewport"[@@noalloc
                                                                    ]
let gl_color_buffer_bit = 16384
let gl_depth_buffer_bit = 256
let gl_accum_buffer_bit = 512
let gl_stencil_buffer_bit = 1024
external clear : int -> unit = "TglClear"[@@noalloc ]
external getUniformLocation :
  program:programT -> name:string -> uniformT = "TglGetUniformLocation"
type attribT
external getAttribLocation :
  program:programT -> name:string -> attribT = "TglGetAttribLocation"
external enableVertexAttribArray :
  attribT -> unit = "TglEnableVertexAttribArray"
let gl_byte = 5120
let gl_unsigned_byte = 5121
let gl_short = 5122
let gl_unsigned_short = 5123
let gl_int = 5124
let gl_unsigned_int = 5125
let gl_float = 5126
let gl_double = 5130
external vertexAttribPointer :
  index:attribT ->
    size:int -> typ:int -> normalize:bool -> stride:int -> offset:int -> unit
    = "TglVertexAttribPointer_bytecode" "TglVertexAttribPointer_native"
external vertexAttribDivisor :
  attribute:attribT -> divisor:int -> unit = "TglVertexAttribDivisor_byte"
let gl_shader_type = 35663
let gl_delete_status = 35712
let gl_compile_status = 35713
let gl_link_status = 35714
let gl_validate_status = 35715
let gl_info_log_length = 35716
let gl_attached_shaders = 35717
let gl_active_uniforms = 35718
let gl_active_uniform_max_length = 35719
let gl_shader_source_length = 35720
let gl_active_attributes = 35721
let gl_active_attribute_max_length = 35722
external getProgramiv :
  program:programT -> pname:int -> int = "TglGetProgramiv"[@@noalloc ]
external getShaderiv : shader:shaderT -> pname:int -> int = "TglGetShaderiv"
[@@noalloc ]
external getShaderInfoLog : shaderT -> string = "TglGetShaderInfoLog"
external getProgramInfoLog : programT -> string = "TglGetProgramInfoLog"
external getShaderSource : shaderT -> string = "TglGetShaderSource"
let gl_points = 0
let gl_lines = 1
let gl_line_loop = 2
let gl_line_strip = 3
let gl_triangles = 4
let gl_triangle_strip = 5
let gl_triangle_fan = 6
let gl_quads = 7
let gl_quad_strip = 8
let gl_polygon = 9
external drawArrays :
  mode:int -> first:int -> count:int -> unit = "TglDrawArrays"[@@noalloc ]
external drawElements :
  mode:int -> count:int -> typ:int -> offset:int -> unit = "TglDrawElements"
[@@noalloc ]
external drawElementsInstanced :
  mode:int -> count:int -> type_:int -> indices:int -> primcount:int -> unit
    = "TglDrawElementsInstanced"[@@noalloc ]
external uniformMatrix4fv :
  location:uniformT -> transpose:bool -> value:float array -> unit =
    "TglUniformMatrix4fv"[@@noalloc ]
type imageT =
  {
  width: int ;
  height: int ;
  channels: int ;
  data:
    (int, Bigarray.int8_unsigned_elt, Bigarray.c_layout)
      Bigarray.Array1.t
    }
external soilLoadImage :
  filename:string -> loadOption:int -> imageT option = "load_image"
external soilLoadImageFromMemory :
  data:string -> loadOption:int -> imageT option = "load_image_from_memory"
external checkFramebufferStatus : unit -> int = "TglCheckFramebufferStatus"
[@@noalloc ]
type renderBufferT
external genRenderbuffers : unit -> renderBufferT = "TglGenRenderbuffers"
[@@noalloc ]
external bindRenderbuffer : renderBufferT -> unit = "TglBindRenderbuffer"
[@@noalloc ]
external bindDefaultRenderbuffer :
  unit -> unit = "TglBindDefaultRenderbuffer"[@@noalloc ]
external renderbufferStorage : int -> int -> unit = "TglRenderbufferStorage"
[@@noalloc ]
external framebufferRenderbuffer :
  renderBufferT -> unit = "TglFramebufferRenderbuffer"[@@noalloc ]

#end
